File: core/src/main/javascript/msg/ErrorHeader.js
/**
 * Copyright (c) 2012-2018 Netflix, Inc.  All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * <p>The error data is represented as
 * {@code
 * errordata = {
 *   "#mandatory" : [ "messageid", "errorcode" ],
 *   "timestamp" : "int64(0,2^53^)",
 *   "messageid" : "int64(0,-)",
 *   "errorcode" : "int32(0,-)",
 *   "internalcode" : "int32(0,-)",
 *   "errormsg" : "string",
 *   "usermsg" : "string",
 * }} where:
 * <ul>
 * <li>{@code timestamp} is the sender time when the header is created in seconds since the UNIX epoch</li>
 * <li>{@code messageid} is the message ID</li>
 * <li>{@code errorcode} is the error code</li>
 * <li>{@code internalcode} is an service-specific error code</li>
 * <li>{@code errormsg} is a developer-consumable error message</li>
 * <li>{@code usermsg} is a user-consumable localized error message</li>
 * </ul></p>
 *
 * @author Wesley Miaw <wmiaw@netflix.com>
 */
(function(require, module) {
	"use strict";
	
	var Class = require('../util/Class.js');
	var MslConstants = require('../MslConstants.js');
	var MslInternalException = require('../MslInternalException.js');
	var MslMessageException = require('../MslMessageException.js');
	var MslError = require('../MslError.js');
	var AsyncExecutor = require('../util/AsyncExecutor.js');
	var MslEncoderException = require('../io/MslEncoderException.js');
	var MslEntityAuthException = require('../MslEntityAuthException.js');
	var MslCryptoException = require('../MslCryptoException.js');
	var Header = require('../msg/Header.js');
	var MslException = require('../MslException.js');
	var MslEncodingException = require('../MslEncodingException.js');
	var Base64 = require('../util/Base64.js');
    
    /** Milliseconds per second. */
    var MILLISECONDS_PER_SECOND = 1000;

    // Message error data.
    /**
     * Key timestamp.
     * @const
     * @type {string}
     */
    var KEY_TIMESTAMP = "timestamp";
    /**
     * Key message ID.
     * @const
     * @type {string}
     */
    var KEY_MESSAGE_ID = "messageid";
    /**
     * Key error code.
     * @const
     * @type {string}
     */
    var KEY_ERROR_CODE = "errorcode";
    /**
     * Key internal code.
     * @const
     * @type {string}
     */
    var KEY_INTERNAL_CODE = "internalcode";
    /**
     * Key error message.
     * @const
     * @type {string}
     */
    var KEY_ERROR_MESSAGE = "errormsg";
    /**
     * Key user message.
     * @const
     * @type {string}
     */
    var KEY_USER_MESSAGE = "usermsg";

    /**
     * Create a new error data container object.
     *
     * @param {Uint8Array} errordata raw error data.
     * @param {number} timestampSeconds creation timestamp in seconds since the epoch.
     * @constructor
     */
    function CreationData(errordata, timestampSeconds) {
        this.errordata = errordata;
        this.timestampSeconds = timestampSeconds;
    }

    var ErrorHeader = module.exports = Class.create({
        /**
         * <p>Construct a new error header with the provided error data.</p>
         *
         * @param {MslContext} ctx MSL context.
         * @param {EntityAuthenticationData} entityAuthData the entity authentication data.
         * @param {number} messageId the message ID.
         * @param {MslConstants.ResponseCode} errorCode the error code.
         * @param {number} internalCode the internal code. Negative to indicate no code.
         * @param {?string} errorMsg the error message. May be null.
         * @param {?string} userMsg the user message. May be null.
         * @param {?CreationData} creationData optional creation data.
         * @throws MslEncodingException if there is an error encoding the JSON
         *         data.
         * @throws MslCryptoException if there is an error encrypting or signing
         *         the message.
         * @throws MslEntityAuthException if there is an error with the entity
         *         authentication data.
         * @throws MslMessageException if no entity authentication data is
         *         provided.
         */
        init: function init(ctx, entityAuthData, messageId, errorCode, internalCode, errorMsg, userMsg, creationData) {
            if (internalCode < 0)
                internalCode = -1;

            // Message ID must be within range.
            if (messageId < 0 || messageId > MslConstants.MAX_LONG_VALUE)
                throw new MslInternalException("Message ID " + messageId + " is out of range.");

            // Message entity must be provided.
            if (!entityAuthData)
                throw new MslMessageException(MslError.MESSAGE_ENTITY_NOT_FOUND);

            // Construct the error data.
            var timestampSeconds, errordata;
            if (!creationData) {
                timestampSeconds = parseInt(ctx.getTime() / MILLISECONDS_PER_SECOND);

                // Construct the error data.
                var encoder = ctx.getMslEncoderFactory();
                errordata = encoder.createObject();
                errordata.put(KEY_TIMESTAMP, timestampSeconds);
                errordata.put(KEY_MESSAGE_ID, messageId);
                errordata.put(KEY_ERROR_CODE, errorCode);
                if (internalCode > 0) errordata.put(KEY_INTERNAL_CODE, internalCode);
                if (errorMsg) errordata.put(KEY_ERROR_MESSAGE, errorMsg);
                if (userMsg) errordata.put(KEY_USER_MESSAGE, userMsg);
            } else {
                timestampSeconds = creationData.timestampSeconds;
                errordata = creationData.errordata;
            }

            // The properties.
            var props = {
                /**
                 * MSL context.
                 * @type {MslContext}
                 */
                ctx: { value: ctx, writable: false, enumerable: false, configurable: false },
                /**
                 * Entity authentication data.
                 * @type {EntityAuthenticationData}
                 */
                entityAuthenticationData: { value: entityAuthData, writable: false, configurable: false },
                /**
                 * Timestamp in seconds since the epoch.
                 * @type {?number}
                 */
                timestampSeconds: { value: timestampSeconds, writable: false, enumerable: false, configurable: false },
                /**
                 * Message ID.
                 * @type {number}
                 */
                messageId: { value: messageId, writable: false, configurable: false },
                /**
                 * Error code.
                 * @type {MslConstants.ResponseCode}
                 */
                errorCode: { value: errorCode, writable: false, configurable: false },
                /**
                 * Internal code.
                 * @type {number}
                 */
                internalCode: { value: internalCode, writable: false, configurable: false },
                /**
                 * Error message.
                 * @type {?string}
                 */
                errorMessage: { value: errorMsg, writable: false, configurable: false },
                /**
                 * User message.
                 * @type {?string}
                 */
                userMessage: { value: userMsg, writable: false, configurable: false },
                /**
                 * Error data.
                 * @type {MslObject}
                 */
                errordata: { value: errordata, writable: false, enumerable: false, configurable: false },
                /**
                 * Cached encodings.
                 * @type {Object<MslEncoderFormat,Uint8Array>}
                 */
                encodings: { value: {}, writable: false, enumerable: false, configurable: false },
            };
            Object.defineProperties(this, props);
        },

        /**
        * @return {Date} gets the timestamp. May be null.
        */
        get timestamp() {
            if (this.timestampSeconds !== null)
                return new Date(this.timestampSeconds * MILLISECONDS_PER_SECOND);
            return null;
        },
        
        /** @inheritDoc */
        toMslEncoding: function toMslEncoding(encoder, format, callback) {
            var self = this;
            
            AsyncExecutor(callback, function() {
                // Return any cached encoding.
                if (this.encodings[format])
                    return this.encodings[format];
                
                // Create the crypto context.
                var scheme = this.entityAuthenticationData.scheme;
                var factory = this.ctx.getEntityAuthenticationFactory(scheme);
                if (!factory)
                    throw new MslEncoderException("No entity authentication factory found for entity.");
                var cryptoContext;
                try {
                    cryptoContext = factory.getCryptoContext(this.ctx, this.entityAuthenticationData);
                } catch (e) {
                    if (e instanceof MslEntityAuthException || e instanceof MslCryptoException)
                        throw new MslEncoderException("Error creating the entity crypto context.", e);
                    throw e;
                }
                
                // Encrypt and sign the error data.
                encoder.encodeObject(this.errordata, format, {
                	result: function(plaintext) {
		                cryptoContext.encrypt(plaintext, encoder, format, {
		                    result: function(ciphertext) {
		                        cryptoContext.sign(ciphertext, encoder, format, {
		                            result: function(signature) {
		                                AsyncExecutor(callback, function() {
		                                    // Create the encoding.
		                                    var header = encoder.createObject();
		                                    header.put(Header.KEY_ENTITY_AUTHENTICATION_DATA, this.entityAuthenticationData);
		                                    header.put(Header.KEY_ERRORDATA, ciphertext);
		                                    header.put(Header.KEY_SIGNATURE, signature);
		                                    encoder.encodeObject(header, format, {
		                                    	result: function(encoding) {
		                                    		AsyncExecutor(callback, function() {
		                                    			// Cache and return the encoding.
		                                    			this.encodings[format] = encoding;
		                                    			return encoding;
		                                    		}, self);
		                                    	},
		                                    	error: callback.error,
		                                    });
		                                }, self);
		                            },
		                            error: function(e) {
		                                if (e instanceof MslCryptoException)
		                                    e = new MslEncoderException("Error signing the error data.", e);
		                                callback.error(e);
		                            }
		                        });
		                    },
		                    error: function(e) {
		                        if (e instanceof MslCryptoException)
		                            e = new MslEncoderException("Error signing the error data.", e);
		                        callback.error(e);
		                    }
		                });
                	},
                	error: callback.error,
                });
            }, self);
        },
    });

    /**
     * <p>Construct a new error header with the provided error data.</p>
     *
     * @param {MslContext} ctx MSL context.
     * @param {EntityAuthenticationData} entityAuthData the entity authentication data.
     * @param {number} messageId the message ID.
     * @param {MslConstants.ResponseCode} errorCode the error code.
     * @param {number} internalCode the internal code. Negative to indicate no code.
     * @param {?string} errorMsg the error message. May be null.
     * @param {?string} userMsg the user message. May be null.
     * @param {{result: function(ErrorHeader), error: function(Error)}}
     *        callback the callback functions that will receive the error
     *        header or any thrown exceptions.
     * @throws MslEncodingException if there is an error encoding the JSON
     *         data.
     * @throws MslCryptoException if there is an error encrypting or signing
     *         the message.
     * @throws MslEntityAuthException if there is an error with the entity
     *         authentication data.
     */
    var ErrorHeader$create = function ErrorHeader$create(ctx, entityAuthData, messageId, errorCode, internalCode, errorMsg, userMsg, callback) {
        AsyncExecutor(callback, function() {
            return new ErrorHeader(ctx, entityAuthData, messageId, errorCode, internalCode, errorMsg, userMsg, null);
        });
    };

    /**
     * <p>Construct a new error header from the provided MSL object.</p>
     *
     * @param {MslContext} ctx MSL context.
     * @param {Uint8Array} errordataBytes error data MSL encoding.
     * @param {EntityAuthenticationData} entityAuthData the entity authentication data.
     * @param {Uint8Array} signature the header signature.
     * @param {{result: function(ErrorHeader), error: function(Error)}}
     *        callback the callback functions that will receive the error
     *        header or any thrown exceptions.
     * @throws MslEncodingException if there is an error parsing the JSON.
     * @throws MslCryptoException if there is an error decrypting or verifying
     *         the header.
     * @throws MslEntityAuthException if the entity authentication data is not
     *         supported or erroneous.
     * @throws MslMessageException if there is no entity authentication data
     *         (null) or the error data is missing or the message ID is
     *         negative or the internal code is negative.
     */
    var ErrorHeader$parse = function ErrorHeader$parse(ctx, errordataBytes, entityAuthData, signature, callback) {
        AsyncExecutor(callback, function() {
            var encoder = ctx.getMslEncoderFactory();
            
            // Validate the entity authentication data.
            if (!entityAuthData)
                throw new MslMessageException(MslError.MESSAGE_ENTITY_NOT_FOUND);

            // Grab the entity crypto context.
            var cryptoContext;
            try {
                var scheme = entityAuthData.scheme;
                var factory = ctx.getEntityAuthenticationFactory(scheme);
                if (!factory)
                    throw new MslEntityAuthException(MslError.ENTITYAUTH_FACTORY_NOT_FOUND, scheme);
                cryptoContext = factory.getCryptoContext(ctx, entityAuthData);
            } catch (e) {
                if (e instanceof MslException)
                    e.setEntityAuthenticationData(entityAuthData);
                throw e;
            }
            
            // Verify and decrypt the error data.
            cryptoContext.verify(errordataBytes, signature, encoder, {
                result: function(verified) {
                    AsyncExecutor(callback, function() {
                        if (!verified)
                            throw new MslCryptoException(MslError.MESSAGE_VERIFICATION_FAILED).setEntityAuthenticationData(entityAuthData);
                        cryptoContext.decrypt(errordataBytes, encoder, {
                            result: function(plaintext) {
                                AsyncExecutor(callback, function() {
                                    var errordata, messageId;
                                    try {
                                        errordata = encoder.parseObject(plaintext);
                                        messageId = errordata.getLong(KEY_MESSAGE_ID);
                                        if (messageId < 0 || messageId > MslConstants.MAX_LONG_VALUE)
                                            throw new MslMessageException(MslError.MESSAGE_ID_OUT_OF_RANGE, "errordata " + errordata).setEntityAuthenticationData(entityAuthData);
                                    } catch (e) {
                                        if (e instanceof MslEncoderException)
                                            throw new MslEncodingException(MslError.MSL_PARSE_ERROR, "errordata " + Base64.encode(plaintext), e).setEntityAuthenticationData(entityAuthData);
                                        throw e;
                                    }
                                    
                                    var timestampSeconds, errorCode, internalCode, errorMsg, userMsg;
                                    try {
                                        timestampSeconds = (errordata.has(KEY_TIMESTAMP)) ? errordata.getLong(KEY_TIMESTAMP) : null;
                                        
                                        // If we do not recognize the error code then default to fail.
                                        errorCode = errordata.getInt(KEY_ERROR_CODE);
                                        var recognized = false;
                                        for (var code in MslConstants.ResponseCode) {
                                            if (MslConstants.ResponseCode[code] == errorCode) {
                                                recognized = true;
                                                break;
                                            }
                                        }
                                        if (!recognized)
                                            errorCode = MslConstants.ResponseCode.FAIL;
                                        
                                        if (errordata.has(KEY_INTERNAL_CODE)) {
                                            internalCode = errordata.getInt(KEY_INTERNAL_CODE);
                                            if (internalCode < 0)
                                                throw new MslMessageException(MslError.INTERNAL_CODE_NEGATIVE, "errordata " + errordata).setEntityAuthenticationData(entityAuthData).setMessageId(messageId);
                                        } else {
                                            internalCode = -1;
                                        }
                                        errorMsg = errordata.optString(KEY_ERROR_MESSAGE, null);
                                        userMsg = errordata.optString(KEY_USER_MESSAGE, null);
                                    } catch (e) {
                                        if (e instanceof MslEncoderException)
                                            throw new MslEncodingException(MslError.MSL_PARSE_ERROR, "errordata " + errordata, e).setEntityAuthenticationData(entityAuthData).setMessageId(messageId);
                                        throw e;
                                    }

                                    // Return the error header.
                                    var creationData = new CreationData(errordata, timestampSeconds);
                                    return new ErrorHeader(ctx, entityAuthData, messageId, errorCode, internalCode, errorMsg, userMsg, creationData);
                                });
                            },
                            error: function(e) {
                                AsyncExecutor(callback, function() {
                                    if (e instanceof MslException)
                                        e.setEntityAuthenticationData(entityAuthData);
                                    throw e;
                                });
                            }
                        });
                    });
                },
                error: function(e) {
                    AsyncExecutor(callback, function() {
                        if (e instanceof MslException)
                            e.setEntityAuthenticationData(entityAuthData);
                        throw e;
                    });
                }
            });
        });
    };
    
    // Exports.
    module.exports.create = ErrorHeader$create;
    module.exports.parse = ErrorHeader$parse;
})(require, (typeof module !== 'undefined') ? module : mkmodule('ErrorHeader'));

File: core/src/main/cpp/msg/ErrorHeader.cpp
/**
 * Copyright (c) 2016-2017 Netflix, Inc.  All rights reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <msg/ErrorHeader.h>
#include <crypto/ICryptoContext.h>
#include <entityauth/EntityAuthenticationData.h>
#include <entityauth/EntityAuthenticationFactory.h>
#include <entityauth/EntityAuthenticationScheme.h>
#include <msg/HeaderKeys.h>
#include <io/MslEncoderFactory.h>
#include <MslCryptoException.h>
#include <MslEncodingException.h>
#include <MslEntityAuthException.h>
#include <MslError.h>
#include <MslInternalException.h>
#include <MslMessageException.h>
#include <util/MslContext.h>
#include <util/MslUtils.h>
#include <sstream>

using namespace std;
using namespace netflix::msl;
using namespace netflix::msl::crypto;
using namespace netflix::msl::entityauth;
using namespace netflix::msl::io;
using namespace netflix::msl::util;

namespace netflix {
namespace msl {
namespace msg {

namespace {
/** Milliseconds per second. */
const int64_t MILLISECONDS_PER_SECOND = 1000;

// Message error data.
/** Key timestamp. */
const string KEY_TIMESTAMP = "timestamp";
/** Key message ID. */
const string KEY_MESSAGE_ID = "messageid";
/** Key error code. */
const string KEY_ERROR_CODE = "errorcode";
/** Key internal code. */
const string KEY_INTERNAL_CODE = "internalcode";
/** Key error message. */
const string KEY_ERROR_MESSAGE = "errormsg";
/** Key user message. */
const string KEY_USER_MESSAGE = "usermsg";

} // namespace anonymous

ErrorHeader::ErrorHeader(shared_ptr<MslContext> ctx, shared_ptr<EntityAuthenticationData> entityAuthData,
        int64_t messageId, const MslConstants::ResponseCode& errorCode,
        int32_t internalCode, const string& errorMsg, const string& userMsg)
    : ctx_(ctx)
    , entityAuthData_(entityAuthData)
    , timestamp_(ctx->getTime() / MILLISECONDS_PER_SECOND)
    , messageId_(messageId)
    , errorCode_(errorCode)
    , internalCode_((internalCode >= 0) ? internalCode : -1)
    , errorMsg_(errorMsg)
    , userMsg_(userMsg)
{
    // Message ID must be within range.
    if (messageId < 0 || messageId > MslConstants::MAX_LONG_VALUE) {
        stringstream ss;
        ss << "Message ID " << messageId << " is out of range.";
        throw MslInternalException(ss.str());
    }

    // Message entity must be provided.
    if (!entityAuthData)
        throw MslMessageException(MslError::MESSAGE_ENTITY_NOT_FOUND);

    // Construct the error data.
    shared_ptr<MslEncoderFactory> encoder = ctx->getMslEncoderFactory();
    errordata_ = encoder->createObject();
    errordata_->put(KEY_TIMESTAMP, timestamp_);
    errordata_->put(KEY_MESSAGE_ID, messageId_);
    errordata_->put(KEY_ERROR_CODE, errorCode_.value());
    if (internalCode_ > 0) errordata_->put(KEY_INTERNAL_CODE, internalCode_);
    if (!errorMsg_.empty()) errordata_->put(KEY_ERROR_MESSAGE, errorMsg_);
    if (!userMsg_.empty()) errordata_->put(KEY_USER_MESSAGE, userMsg_);
}

ErrorHeader::ErrorHeader(shared_ptr<MslContext> ctx, shared_ptr<ByteArray> errordataBytes,
        shared_ptr<EntityAuthenticationData> entityAuthData, shared_ptr<ByteArray> signature)
{
    ctx_ = ctx;

    shared_ptr<MslEncoderFactory> encoder = ctx->getMslEncoderFactory();

    shared_ptr<ByteArray> plaintext;
    try {
        // Validate the entity authentication data.
        entityAuthData_ = entityAuthData;
        if (!entityAuthData)
            throw MslMessageException(MslError::MESSAGE_ENTITY_NOT_FOUND);

        // Grab the entity crypto context.
        const EntityAuthenticationScheme scheme = entityAuthData->getScheme();
        shared_ptr<EntityAuthenticationFactory> factory = ctx->getEntityAuthenticationFactory(scheme);
        if (!factory)
            throw MslEntityAuthException(MslError::ENTITYAUTH_FACTORY_NOT_FOUND, scheme.name());
        shared_ptr<ICryptoContext> cryptoContext = factory->getCryptoContext(ctx, entityAuthData);

        // Verify and decrypt the error data.
        if (!cryptoContext->verify(errordataBytes, signature, encoder))
            throw MslCryptoException(MslError::MESSAGE_VERIFICATION_FAILED).setEntityAuthenticationData(entityAuthData);
        plaintext = cryptoContext->decrypt(errordataBytes, encoder);
    } catch (MslCryptoException& e) {
        e.setEntityAuthenticationData(entityAuthData);
        throw e;
    } catch (MslEntityAuthException& e) {
        e.setEntityAuthenticationData(entityAuthData);
        throw e;
    }

    try {
        errordata_ = encoder->parseObject(plaintext);
        messageId_ = errordata_->getLong(KEY_MESSAGE_ID);
        if (messageId_ < 0 || messageId_ > MslConstants::MAX_LONG_VALUE)
            throw MslMessageException(MslError::MESSAGE_ID_OUT_OF_RANGE, "errordata " + errordata_->toString()).setEntityAuthenticationData(entityAuthData_);
    } catch (const MslEncoderException& e) {
        throw MslEncodingException(MslError::MSL_PARSE_ERROR, "errordata " + *Base64::encode(plaintext), e).setEntityAuthenticationData(entityAuthData_);
    }

    try {
        timestamp_ = (errordata_->has(KEY_TIMESTAMP)) ? errordata_->getLong(KEY_TIMESTAMP) : -1;

        // If we do not recognize the error code then default to fail.
        MslConstants::ResponseCode code = MslConstants::ResponseCode::FAIL;
        try {
            code = MslConstants::ResponseCode::valueOf(errordata_->getInt(KEY_ERROR_CODE));
        } catch (const IllegalArgumentException& e) {
            code = MslConstants::ResponseCode::FAIL;
        }
        errorCode_ = code;

        if (errordata_->has(KEY_INTERNAL_CODE)) {
            internalCode_ = errordata_->getInt(KEY_INTERNAL_CODE);
            if (internalCode_ < 0)
                throw MslMessageException(MslError::INTERNAL_CODE_NEGATIVE, "errordata " + errordata_->toString()).setEntityAuthenticationData(entityAuthData_).setMessageId(messageId_);
        } else {
            internalCode_ = -1;
        }
        errorMsg_ = errordata_->optString(KEY_ERROR_MESSAGE, string());
        userMsg_ = errordata_->optString(KEY_USER_MESSAGE, string());
    } catch (const MslEncoderException& e) {
        throw MslEncodingException(MslError::MSL_PARSE_ERROR, "errordata " + errordata_->toString(), e).setEntityAuthenticationData(entityAuthData_).setMessageId(messageId_);
    }
}

shared_ptr<Date> ErrorHeader::getTimestamp() const
{
    return (timestamp_ != -1) ? make_shared<Date>(timestamp_ * MILLISECONDS_PER_SECOND) : shared_ptr<Date>();
}

shared_ptr<ByteArray> ErrorHeader::toMslEncoding(shared_ptr<io::MslEncoderFactory> encoder, const io::MslEncoderFormat& format) const
{
    // Return any cached encoding.
    std::map<io::MslEncoderFormat, shared_ptr<ByteArray>>::const_iterator it = encodings_.find(format);
    if (it != encodings_.end())
        return it->second;

    // Create the crypto context.
    const EntityAuthenticationScheme scheme = entityAuthData_->getScheme();
    shared_ptr<EntityAuthenticationFactory> factory = ctx_->getEntityAuthenticationFactory(scheme);
    if (!factory)
        throw MslEncoderException("No entity authentication factory found for entity.");
    shared_ptr<ICryptoContext> cryptoContext;
    try {
        cryptoContext = factory->getCryptoContext(ctx_, entityAuthData_);
    } catch (const MslEntityAuthException& e) {
        throw MslEncoderException("Error creating the entity crypto context.", e);
    } catch (const MslCryptoException& e) {
        throw MslEncoderException("Error creating the entity crypto context.", e);
    }

    // Encrypt and sign the error data.
    shared_ptr<ByteArray> plaintext = encoder->encodeObject(errordata_, format);
    shared_ptr<ByteArray> ciphertext;
    try {
    	ciphertext = cryptoContext->encrypt(plaintext, encoder, format);
    } catch (const MslCryptoException& e) {
    	throw MslEncoderException("Error encrypting the error data.", e);
    }
    shared_ptr<ByteArray> signature;
    try {
    	signature = cryptoContext->sign(ciphertext, encoder, format);
    } catch (const MslCryptoException& e) {
    	throw MslEncoderException("Error signing the error data.", e);
    }

    // Create the encoding.
    shared_ptr<MslObject> header = encoder->createObject();
    header->put<shared_ptr<MslEncodable>>(HeaderKeys::KEY_ENTITY_AUTHENTICATION_DATA, entityAuthData_);
    header->put(HeaderKeys::KEY_ERRORDATA, ciphertext);
    header->put(HeaderKeys::KEY_SIGNATURE, signature);
    shared_ptr<ByteArray> encoding = encoder->encodeObject(header, format);

    // Cache and return the encoding.
    encodings_.insert(make_pair(format, encoding));
    return encoding;
}

// In this translation unit, make comparisons of shared_ptr defer to the
// underlying type's operator==.
template <typename T>
bool operator==(const shared_ptr<T>& a, const shared_ptr<T>& b)
{
    if (!a && !b)
        return true;
    if ((!a && b) || (a && !b))
        return false;
    return *a == *b;
}

bool ErrorHeader::equals(shared_ptr<const Header> obj) const
{
	if (!obj) return false;
	if (this == obj.get()) return true;
    if (!instanceof<const ErrorHeader>(obj)) return false;
    shared_ptr<const ErrorHeader> that = dynamic_pointer_cast<const ErrorHeader>(obj);
    // Don't need to check shared_ptr for null, since templated operator==() above does this.
    return (entityAuthData_ == that->entityAuthData_) &&
			(timestamp_ == that->timestamp_) &&
			(messageId_ == that->messageId_) &&
			(errorCode_ == that->errorCode_) &&
			(internalCode_ == that->internalCode_) &&
			(errorMsg_ == that->errorMsg_) &&
			(userMsg_ == that->userMsg_);
}

bool operator==(const ErrorHeader& a, const ErrorHeader& b)
{
	shared_ptr<const ErrorHeader> ap(&a, &MslUtils::nullDeleter<ErrorHeader>);
	shared_ptr<const ErrorHeader> bp(&b, &MslUtils::nullDeleter<ErrorHeader>);
	return ap->equals(bp);
}

ostream& operator<<(ostream& os, const ErrorHeader& /*header*/)
{
	// FIXME
	return os << "placeholder";
}

ostream& operator<<(ostream& os, shared_ptr<ErrorHeader> header)
{
	return os << *header;
}

}}} // namespace netflix::msl::msg
